### DISCUSSION TITLE
Systemd Service

### DISCUSSION THREAD

**Discussion by djarbz** [Chia-Network/chia-blockchain#3987] [Support]
Is there a better way to run a wallet/node as a service?
I have this working for now, but seems clunky?

Does the daemon automatically restart the child processes if they fail?

Here is my service file
```
# chia-node.service

[Unit]
Description="Chia Blockchain Node service"
After=network.target

[Service]
Type=simple
RestartSec=1s
Restart=always
User=chia
Group=chia
ExecStop=/usr/local/bin/chia stop wallet
ExecStopPost=kill -9 $(pidof chia_daemon)
ExecStartPost=/usr/local/bin/chia start wallet
ExecStart=/usr/local/bin/chia run_daemon

[Install]
WantedBy=multi-user.target
```

And here is what it spawns

```
ps | grep chia
chia     52281  0.0  2.2 193792 46688 ?        SLsl 14:21   0:03 chia_daemon
chia     52714  0.0  2.7 948284 57440 ?        Sl   14:21   0:04  \_ chia_wallet
chia     52715  0.0  8.6 854952 181176 ?       Sl   14:21   0:45  \_ chia_full_node
chia     52958  0.0  7.6 842152 159596 ?       R    14:25   4:41      \_ chia_full_node
```

**Comment by jurepetrovic** [comment]
How does this work for you? 
How do you monitor status? 

I'm looking for the same thing, but I was going into the direction of using pm2 process manager
(http://pm2.io) since it supports restarts and is more aligned with node apps. Any thoughts on that? 

Thanks, 
Jure

**Reply by djarbz** [reply]
I don't work with Node, so I've never heard of PM2.
This is a Python app, I don't recall seeing any Node runtimes.

Basically this service file will start the Chia Daemon and monitor that for crashes and whatnot.
The Start Post option will start whatever child services you want running, in this case, I was running a wallet which requires the Node as well.

I've since rewritten it to do a ```chia stop all -d``` instead of stopping the wallet and then killing the daemon.

**Comment by jurepetrovic** [comment]
I see. PM2 Is not node exclusive, it can be used for all types and it has quite some features. 
Doesn't matter - I will happilly follow your idea - so thank you! 

One more question.
To run full node, is it enough to start "chia start wallet"? 
What do I need to start in order to get full node, same as what the GUI (electron) app does? 

Thanks, 
Jure

**Reply by djarbz** [reply]
You can do ```chia start node```, this is probably most similar to the GUI.
I haven't used the GUI for this, all command line.

I'm still figuring out what all of the start options actually start, haven't found documentation on them yet.
All will start everything, including the Timelord.
Node will start everything except the Timelord.
Wallet will start the Node and the Wallet.
Farmer will start a Node(?), Farmer, and Harvester.

**Comment by jurepetrovic** [comment]
those are the processes running, when starting GUI. 
![Screenshot from 2021-05-10 15-00-42](https://user-images.githubusercontent.com/35005863/117665135-cf8ee980-b1a2-11eb-8672-c4c0a3aa5c7f.png)

Seems like I need node, waller, farmer and harvester processes.

**Reply by djarbz** [reply]
~~Yes, so I believe if you run ```chia start node```, it should run all of those. But I could be mistaken, you may need to also run ```chia start farmer```.~~

EDIT: See table on root thread, you will need ```chia start farmer```

**Comment by djarbz** [comment]
I did some digging, here is what each start option will actually start.

|                   | Full Node | Wallet | Farmer | Harveser | Timelord | Timelord Launcher | Introducer | Full Node Simulator |
|-------------------|-----------|--------|--------|----------|----------|-------------------|------------|---------------------|
| all               | X         | X      | X      | X        | X        | X                 |            |                     |
| node              | X         |        |        |          |          |                   |            |                     |
| harvester         |           |        |        | X        |          |                   |            |                     |
| farmer            | X         | X      | X      | X        |          |                   |            |                     |
| farmer-no-wallet  | X         |        | X      | X        |          |                   |            |                     |
| farmer-only       |           |        | X      |          |          |                   |            |                     |
| timelord          | X         |        |        |          | X        | X                 |            |                     |
| timelord-only     |           |        |        |          | X        |                   |            |                     |
| timelord-launcher |           |        |        |          |          | X                 |            |                     |
| wallet            | X         | X      |        |          |          |                   |            |                     |
| wallet-only       |           | X      |        |          |          |                   |            |                     |
| introducer        |           |        |        |          |          |                   | X          |                     |
| simulator         |           |        |        |          |          |                   |            | X                   |

**Reply by jurepetrovic** [reply]
These are exactly the processes I need! Thank you very much. 

But what is that run_daemon? What is the purpose? I have to start that anyway, right?

**Reply by djarbz** [reply]
run_daemon is the master service, this is what Systemd can monitor. If you just run ```chia start...``` it will start the run_daemon anyways. The ```chia``` command is just a command line client interface, think of the run_daemon as the server.

**Comment by jurepetrovic** [comment]
Ok, I've come a bit further. 
It seems that my script from debian package goes into background: 

/usr/lib/chia-blockchain/resources/app.asar.unpacked/daemon/chia start farmer

which causes the systemd to think that it died, restarting in indefinitely.

**Comment by jurepetrovic** [comment]
I think I'm OK now. 
I hope those processes are ok for farming.

![image](https://user-images.githubusercontent.com/35005863/118250612-f309ab00-b4a6-11eb-9d80-ee827a56110d.png)

I changed startup script type to type=forking. Looks like this now: 

```
# chia-node.service

[Unit]
Description="Chia Blockchain Node service"
After=network.target

[Service]
# runs daemonized
Type=forking
RestartSec=15s
Restart=always
User=jure
Group=jure
ExecStart=/usr/lib/chia-blockchain/resources/app.asar.unpacked/daemon/chia start farmer
ExecStop=/usr/lib/chia-blockchain/resources/app.asar.unpacked/daemon/chia stop all -d

[Install]
WantedBy=multi-user.target
```

Thank you all for help!
Jure

**Reply by djarbz** [reply]
Check your logs, I found that when I had it set to type forking that it was constantly restarting.
This is why I had it start the daemon and then post start the actual service.

**Comment by drewwells** [comment]
I have managed to run a harvester with this systemd. Be sure to add forking or systemd will continuously restart chia since it forks. Also change your logs to stdout so you can track logs from journalctl.

```
cat /etc/systemd/system/harvester.service                                    
[Unit]
Description=chia harvester
Wants=network-online.target
After=network.target network-online.target
StartLimitIntervalSec=0

[Service]
Type=forking
Restart=always
RestartSec=1
User={chia user}

# venv/bin must be in the PATH b/c the scripts expect to be able to call chia_harvester on cli
ExecStart=/usr/bin/env PATH=/path/to/venv/bin:$PATH chia start harvester -r
ExecStop=/usr/bin/env PATH=/path/to/venv/bin:$PATH chia stop all -d

[Install]
WantedBy=multi-user.target
```

**Reply by mcybz** [reply]
This looks good. However is it an issue that we dont run it in a `venv` and instead link directly? Guess I was concerned might be subtle issues not running `source chia-blockchain/activate` first.

**Reply by drewwells** [reply]
That appears to just change your path. The binaries shell out to each other instead so you’ll get errors without the path changes.

**Reply by polarn** [reply]
@drewwells that systemd file works perfectly, thanks! How does one change to stdout logging?

Edit: found it, I was looking under the harvester options only but it seems to work to change the farmer logging options. So for others that are lost, check `farmer.logging.log_stdout` and set it to true. :)

**Reply by drewwells** [reply]
https://github.com/Chia-Network/chia-blockchain/blob/203232e429ca06cd6ec0fb23921b02c675d43dd5/chia/util/initial-config.yaml#L110

**Reply by djarbz** [reply]
When I tried a forking service I found that it kept restarting because I didn't have a PID file to monitor.

I don't use the virtual environment because I have a single purpose system, so CHIA is the only thing running on this machine.

**Reply by drewwells** [reply]
Forking checks that the process exits, not that it is running. https://www.freedesktop.org/software/systemd/man/systemd.service.html

Check for syntax errors, my process has been running for two days without a restart. Notice that it figured out the PID of the child process without configuration.

```
-> % systemctl status harvester
● harvester.service - chia harvester
     Loaded: loaded (/etc/systemd/system/harvester.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2021-05-28 19:47:09 CDT; 2 days ago
   Main PID: 2455 (chia_daemon)
      Tasks: 9 (limit: 38331)
     Memory: 72.9M
     CGroup: /system.slice/harvester.service
             ├─2455 chia_daemon
             └─2746 chia_harvester
```

**Reply by shokinn** [reply]
@drewwells 
It's maybe a bit cleaner to use the Environment option of systemd instead to integrate it to the exec comand :)

And thanks for the script 👍 

```systemd
[Unit]
Description=chia harvester
Wants=network-online.target
After=network.target network-online.target
StartLimitIntervalSec=0

[Service]
Type=forking
Restart=always
RestartSec=1
User={chia user}

# venv/bin must be in the PATH b/c the scripts expect to be able to call chia_harvester on cli
Environment=PATH=/path/to/venv/bin:${PATH}
ExecStart=/usr/bin/env chia start harvester -r
ExecStop=/usr/bin/env chia stop all -d

[Install]
WantedBy=multi-user.target
```

**Comment by qq597353** [comment]
非常有帮助,感谢🙏

**Comment by BrandtH22** [comment]
Hey @djarbz , the upcoming release version (2.1.2) should contain an official systemd service file:
https://github.com/Chia-Network/chia-blockchain/commit/dcea5363c695620ec358f95fc95124b225095910

Keep an eye open for this release and let us know if any questions arise.
The best place to reach our support team is on Discord (https://discord.gg/chia).
